src/Container.php

```diff
@@ -136,65 +149,79 @@ public function has(string $id): bool
-                    throw $e;
-                }
-                throw new BuildingException($id, $e, $this->definitions->getBuildStack(), $e);
+        try {
+            $this->instances[$id] = $this->build($id);
+        } catch (NotFoundException $exception) {
+            // Fast path: if the exception ID matches the requested ID, no need to modify stack.
+            if ($exception->getId() === $id) {
+                // Try delegates before giving up.
+                return $this->delegates->has($id) ? $this->delegates->get($id) : throw $exception;
+            }
+
+            // Add current ID to build stack for better error reporting.
+            $buildStack = $exception->getBuildStack();
+            array_unshift($buildStack, $id);
+            throw new NotFoundException($exception->getId(), $buildStack);
+        } catch (NotFoundExceptionInterface $exception) {
+            // Try delegates before giving up
+            return $this->delegates->has($id) ? $this->delegates->get($id) : throw new NotFoundException($id, [$id], previous: $exception);
+        } catch (ContainerExceptionInterface $e) {
+            if (!$e instanceof InvalidConfigException) {
+                throw $e;
             }
+            throw new BuildingException($id, $e, $this->definitions->getBuildStack(), $e);
+        } catch (Throwable $e) {
+            throw new BuildingException($id, $e, $this->definitions->getBuildStack(), $e);
         }
 
+        // Handle StateResetter for newly built instances.
         if ($id === StateResetter::class) {
-            $delegatesResetter = null;
-            if ($this->delegates->has(StateResetter::class)) {
-                $delegatesResetter = $this->delegates->get(StateResetter::class);
-            }
+            return $this->prepareStateResetter($id);
+        }
 
-            /** @var StateResetter $mainResetter */
-            $mainResetter = $this->instances[$id];
+        /** @psalm-suppress MixedReturnStatement */
+        return $this->instances[$id];
+    }
 
-            if ($this->useResettersFromMeta) {
-                /** @var StateResetter[] $resetters */
-                $resetters = [];
-                foreach ($this->resetters as $serviceId => $callback) {
-                    if (isset($this->instances[$serviceId])) {
-                        $resetters[$serviceId] = $callback;
-                    }
-                }
-                if ($delegatesResetter !== null) {
-                    $resetters[] = $delegatesResetter;
-                }
-                $mainResetter->setResetters($resetters);
-            } elseif ($delegatesResetter !== null) {
-                $resetter = new StateResetter($this->get(ContainerInterface::class));
-                $resetter->setResetters([$mainResetter, $delegatesResetter]);
+    /**
+     * @param string $id
+     * @return mixed
+     */
+    private function prepareStateResetter(string $id)
```

Comment:
```suggestion
    private function prepareStateResetter(): StateResetter
```

`$id` is always `StateResetter::class`